File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, a: List[int]) -> List[int]:
        stack = []

        for i in range(len(a)):
            while stack and stack[-1] > 0 and a[i] < 0 and -a[i] > stack[-1]:
                stack.pop()
            if stack and a[i] < 0 and (-a[i] == stack[-1]):
                stack.pop()
                continue
            if stack and a[i] < 0 and (-a[i] < stack[-1]):
                continue
            stack.append(a[i])
        return stack

File: 735-asteroid-collision/asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] as) {
        Stack<Integer> a = new Stack<>();

        for (int i = 0; i < as.length; i++) {
            while (!(a.isEmpty()) && a.peek() > 0 && as[i] < 0 && -as[i] > a.peek()) {
                a.pop();
            }
            if (as[i] < 0) {
                if (!(a.isEmpty()) && a.peek() > 0) {
                    if (-as[i] < a.peek()) {
                        continue;
                    }
                    else {
                        a.pop();
                        continue;
                    }
                }
            }
            a.push(as[i]);
        }
        int[] array = new int[a.size()];
        int i = 0;
        for (int num : a) {
            array[i++] = num;
        }

        // System.out.println(a);
        return array;
    }
}